### PR TITLE
[WIP] Added side bar menu. Added filter button that opens that side bar men…

### PR DESCRIPTION
**Pull request by rootux**
# Implements

* Added side bar menu
* Added filter button that opens that side bar menu
* Added actions to support user interface reducers

🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧
⚠️ THIS PULL REQUEST SHOULD NOT BE MERGED - MORE CHANGES ARE NEEDED ⚠️
🚧🚧🚧🚧🚧🚧 🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧

Pressing on the "Filter" button now opens the men

All the 

The menu should look like this (We should move `TaskFilters` inside the `filter-menu` component.
Screen Shot 2019-10-15 at 22 19 51

The current state
Screen Shot 2019-10-15 at 22 17 14

@netanelben Do you think you can take this from here?

### CODE REVIEW

**Comment by rootux** [comment]
Probably shouldComponentUpdate on the menu sidebar should be checked to see if the new state is different then the internal state to prevent uneccesery rendering

**Comment by rootux** [comment]
Updates - Responsive search bar and filters started the migration to the side bar

Screen Shot 2019-10-16 at 23 54 55
Screen Shot 2019-10-16 at 23 54 50
Screen Shot 2019-10-16 at 23 55 01

**Comment by netanelben** [comment]
@rootux reviewing ...

File: package.json [hunk 1]
```diff
@@ -51,10 +51,12 @@
     "i18next-browser-languagedetector": "2.1.0",
     "i18next-xhr-backend": "1.5.0",
     "immutable": "3.8.2",
+    "lodash.throttle": "4.1.1",
```
Comment (netanelben) [inline package.json]: I'd rather import lodash with all the utilities/features it includes.
Its very powerful tool to have and Im positive we will use it.

**Comment by rootux** [comment]
Having some hebrew Unicode encoding issues
@netanelben 

Screen Shot 2019-10-18 at 11 31 07

**Comment by rootux** [comment]
Another issue - 
• Set a label on side menu filter.
• Clicked on 'X' to remove that label
• Label removed from the main view but NOT from the side menu
Screen Shot 2019-10-18 at 11 32 34

File: package.json [hunk 2]
```diff
@@ -51,10 +51,13 @@
     "i18next-browser-languagedetector": "2.1.0",
     "i18next-xhr-backend": "1.5.0",
     "immutable": "3.8.2",
+    "lodash": "4.17.15",
+    "lodash.throttle": "4.1.1",
```
Comment (netanelben) [inline package.json]: you can remove this line "lodash.throttle

Comment (rootux) [reply package.json]: Thanks. Removed.

File: src/tasks/actions.js [hunk 3]
```diff
@@ -16,7 +16,8 @@ import {
 } from './action-types';
 
 
-export function createTask(task, cb = (t)=>{}) {
+export function createTask(task, user, cb = (t)=>{}) {
```
Comment (netanelben) [inline src/tasks/actions.js:19]: `createTask` the name doesn't comply with the logic..
this func does add users to the task and adds new task to the list ?

Could be much more readable like that:
```JS
export function createTask(task, user, cb = (t) => {}) {
  task.listeners = addUserToListeners(task, user);

  return (dispatch) =>
    taskList.push(task)
      .then(cb)
      .catch(error => dispatch(createTaskError(error)));
}
```

File: src/tasks/reducer.js [hunk 4]
```diff
@@ -24,7 +25,8 @@ export function tasksReducer(state = new TasksState(), {payload, type}) {
       return state.merge({
         deleted: null,
         created: payload,
-        list: state.list.unshift(payload)
+        list: state.list.unshift(payload),
+        labelsPool: state.labelsPool.add(Object.keys(payload.label)[0]),
```
Comment (netanelben) [inline src/tasks/reducer.js]: `state.labelsPool.add(Object.keys(payload.label)[0])` what this line does ?

File: src/tasks/selectors.js [hunk 5]
```diff
@@ -13,6 +15,17 @@ export function buildFilter(auth, type, value) {
   return filters[type](auth, value);
 }
 
+export function _labelsPool(state) {
```
Comment (netanelben) [inline src/tasks/selectors.js]: rather not use `_` prefix, can confuse with lodash

Comment (netanelben) [reply src/tasks/selectors.js]: can easily be arrow function:
```JS
export const labelsPoolSelector = ({ tasks }) => tasks.labelsPool;

export const getLabelsPool = createSelector(labelsPoolSelector, (labelsPool) => labelsPool);
```

File: src/user-interface/index.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import * as userInterfaceActions from './actions';
+
+
```
Comment (netanelben) [inline src/user-interface/index.js]: redundant line-break

File: src/user-interface/reducer.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { Record } from 'immutable';
+import {SET_MENU_OPEN} from './action-types';
+
```
Comment (netanelben) [inline src/user-interface/reducer.js:3]: redundant line-break

File: src/user-interface/reducer.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+export const UserInterfaceState = new Record({
+  isMenuOpen: false,
+});
+
```
Comment (netanelben) [inline src/user-interface/reducer.js:7]: redundant line-break

File: src/user-interface/selectors.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { createSelector } from 'reselect';
+
+
+export function getMenuIsOpenState(state) {
+  return state.userInterface.isMenuOpen;
```
Comment (netanelben) [inline src/user-interface/selectors.js]: Same here and the next function can be arrow functions without the `return` statement.

File: src/utils/browser-utils.js [hunk 10]
```diff
@@ -4,7 +4,11 @@ export function getUrlSearchParams(locationSearch = window.location.search) {
   return locationSearch.substr(1).split('&').reduce(function (q, query) {
     const chunks = query.split('=');
     const key = chunks[0];
-    const value = chunks[1];
+    if(!chunks[1]) {
```
Comment (netanelben) [inline src/utils/browser-utils.js:7]: These functions looks like something we can deprecate... 

`URLSearchParams` API only unavailable for IE-11 maybe we can get around with out it..

If not, we should use polyfills instead - its much more stable.

File: src/tasks/reducer.js (hunk 4 above)
Comment (rootux) [reply src/tasks/reducer.js]: It's adding the labels that are attached to the given task

Comment (rootux) [reply src/tasks/reducer.js]: Oops this was a one year bug. I've fixed it now :)

Comment (rootux) [reply src/tasks/reducer.js]: Great catch

File: src/tasks/selectors.js (hunk 5 above)
Comment (rootux) [reply src/tasks/selectors.js]: Oh god this is so pretty I'm melting here 🥇

File: src/utils/browser-utils.js (hunk 10 above)
Comment (rootux) [reply src/utils/browser-utils.js:7]: Yeah. I've wrote those and I'm sure there are better ways to get browser url query parameters. probably using react route.

Comment (rootux) [reply src/utils/browser-utils.js:7]: I'm not actually using `URLSearchParams` - I believe I've had issues with mobile browsers - can't remember the issue back then - Anyway - Added a Polyfill cause it's 2019 (Hopefully done it the right way) - I also added a few other polyfills as I was freely using other cool latest ES features.
Will revert now to URLSearchParams cause it's probably more trusted then what I did there.

Comment (rootux) [reply src/utils/browser-utils.js:7]: Hmmm `URLSearchParams` is lame and can't handle delete for `label='my'&label='you` it just delete all of `label` so I would need to keep some of my original functions

**Comment by rootux** [comment]
@netanelben Still having issues with Hebrew

Screen Shot 2019-10-18 at 11 31 07

**Pull request merged**